client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ListenerCacheManagerStopTest.java

```diff
@@ -0,0 +1,131 @@
+package org.infinispan.client.hotrod.event;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+import org.infinispan.client.hotrod.test.HotRodClientTestingUtil;
+import org.infinispan.commons.util.Util;
+import org.infinispan.server.hotrod.HotRodServer;
+import org.infinispan.server.hotrod.test.HotRodTestingUtil;
+import org.infinispan.test.AbstractInfinispanTest;
+import org.jboss.byteman.contrib.bmunit.BMNGListener;
+import org.jboss.byteman.contrib.bmunit.BMRule;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import java.net.ServerSocket;
+
+import static org.infinispan.server.hotrod.test.HotRodTestingUtil.hotRodCacheConfiguration;
+import static org.infinispan.test.TestingUtil.extractField;
+import static org.infinispan.test.fwk.TestCacheManagerFactory.createCacheManager;
+
+/**
+ * Test to verify that the listener threads are stopped.
+ *
+ * @author gustavonalle
+ * @since 7.2
+ */
+@Test(groups = "functional", testName = "client.hotrod.event.ListenerStopTest")
+@Listeners(BMNGListener.class)
+public class ListenerCacheManagerStopTest extends AbstractInfinispanTest {
+
+   HotRodServer hotRodServer;
+   RemoteCacheManager remoteCacheManager;
+   RemoteCache<Integer, String> cache;
+
+   @BeforeMethod
+   public void setup() throws IOException {
+      Integer port = getRandomFreePort();
+      hotRodServer = HotRodTestingUtil.startHotRodServer(createCacheManager(hotRodCacheConfiguration()), port);
+      remoteCacheManager = new RemoteCacheManager(
+            new ConfigurationBuilder()
+                  .addServer()
+                  .port(port)
+                  .host(HotRodTestingUtil.host())
+                  .connectionTimeout(3000)
+                  .socketTimeout(3000)
+                  .build()
+      );
+      cache = remoteCacheManager.getCache();
+   }
+
+   @AfterMethod
+   public void tearDown() {
+      HotRodClientTestingUtil.killServers(hotRodServer);
+      HotRodClientTestingUtil.killRemoteCacheManager(remoteCacheManager);
+   }
+
+   /**
+    * Simulates the scenario where SocketInputStream.read throws {@link java.nio.channels.IllegalBlockingModeException}
+    * instead of {@link java.nio.channels.ClosedChannelException}, after a channel is closed.
+    */
+   @Test(invocationCount = 1)
+   @BMRule(name = "replace exception",
```

Comment:
This was based on my debug sessions: I've seen that only sometimes ```read()``` throws ```IllegalBlockingModeException``` when the channel is closed, when it should have thrown ```ClosedChannelException``` or a subclass. This might be related/similar to [1], but I did not dig deeper in the JDK itself.

```read()``` cannot throw ```IllegalBlockingModeException``` directly, it needs to work normally until a certain point, otherwise the test would fail in the setup phase preventing any communication of the client to the server.

[1] https://bugs.openjdk.java.net/browse/JDK-6878250